add support for sprintf

dstroy0 opened the pull request:
now you can use your RF24 object to format a buffer with debugging information

2bndy5 commented:
Well, this came out of nowhere. I tried this a while back, but got stumped on what I thought was reference problems. I have questions:
1. Have you tested this? If so, on what boards/platforms?
2. What are the code-size implications? Does this implementation add to the compile size? We have mainly used `printf()` on ATMega328P and tried to patch it with Arduino cores that support `printf()`. I'm not worried about Linux in this regard.
3. Specifically, what is this targeting? I don't know all the Arduino cores that support `sprintf()` off the top of my head.

2bndy5 commented on RF24.h:
```diff
@@ -598,6 +598,36 @@ class RF24 {
+     * @warning use a buffer of sufficient size, start with a sizeof >= 1024
+     * 
+     * @code
+     * char buffer[1024] = {'\0};
+     * int _len = 0;
+     * RF24 radio(RF24_RADIO_CE_PIN, RF24_RADIO_CS_PIN); 
+     * setup(){
+     * radio.sprintfPrettyDetails(buffer);
+     * _len = strlen(buffer);
+     * Serial.begin(115200);
+     * Serial.print(buffer);
+     * Serial.print(F("strlen = "));
+     * Serial.println(strlen(buffer));
+     * }
```
This sample code block needs proper (4-space) indentation.

Since you copied the docs from the original `printPrettyDetails()`, you should help explain to users how it differs from that (maybe include a link to a C++ reference about using `sprintf()`).

dstroy0 replied:
I fixed the indentation, and added documentation

2bndy5 commented on RF24.h:
```diff
@@ -1854,7 +1884,19 @@ class RF24 {
+     * registers on the same line.  This is useful for printing a group
+     * of related registers on one line.
+     *
+     * @param *out_buffer output buffer
```
you don't need the `*` here

dstroy0 replied:
fixed

2bndy5 commented on RF24.h:
```diff
@@ -1867,6 +1909,18 @@ class RF24 {
+     * registers on the same line.  This is useful for printing a group
+     * of related registers on one line.
+     *
+     * @param ^out_buffer output buffer
```
the `^` seems like a typo, but again you don't need an indication that this is passed by reference (doxygen shows that in the member prototype).

dstroy0 replied:
fixed

2bndy5 commented on RF24.h:
```diff
@@ -1559,8 +1589,8 @@ class RF24 {
      * Request (IRQ) pin active LOW.
      * The following events can be configured:
      * 1. "data sent": This does not mean that the data transmitted was
-     * received, only that the attempt to send it was complete.
-     * 2. "data failed": This means the data being sent was not received. This
+     * recieved, only that the attempt to send it was complete.
+     * 2. "data failed": This means the data being sent was not recieved. This
```
The spelling for the word "received" has already been corrected in #817. Please rebase your master branch on upstream master.

dstroy0 replied:
I'm still trying to figure out how to rebase from the upstream master.

2bndy5 replied:
I checked your branch's commit history, and it looks like the merge commit from #817 is in your branches history already. So, I don't know why this is showing up in the diff. Although, it should be easy to replace all "recieved" with the properly spelled "received".

2bndy5 replied:
I meant that you should re-do the search-and-replace all... Please respect the work that's been done and don't add to it (especially if its repetitive).

dstroy0 replied:
I see, I have been using the client now hopefully correctly.  Thanks for taking the time to explain this stuff to me, I appreciate it.

2bndy5 commented on RF24.cpp:
```diff
@@ -770,19 +801,175 @@ void RF24::printPrettyDetails(void) {
+            (char *)(pgm_read_ptr(&rf24_feature_e_str_P[static_cast<bool>(read_register(FEATURE) & _BV(EN_DYN_ACK)) * 2])),
+            (char *)(pgm_read_ptr(&rf24_feature_e_str_P[static_cast<bool>(read_register(FEATURE) & _BV(EN_ACK_PAY)) * 1])),
+            (char *)(pgm_read_ptr(&rf24_feature_e_str_P[(read_register(DYNPD) && (read_register(FEATURE) &_BV(EN_DPL))) * 1])),
+            static_cast<char>(static_cast<bool>(read_register(EN_AA) & _BV(ENAA_P5)) + 48),
+            static_cast<char>(static_cast<bool>(read_register(EN_AA) & _BV(ENAA_P4)) + 48),
+            static_cast<char>(static_cast<bool>(read_register(EN_AA) & _BV(ENAA_P3)) + 48),
+            static_cast<char>(static_cast<bool>(read_register(EN_AA) & _BV(ENAA_P2)) + 48),
+            static_cast<char>(static_cast<bool>(read_register(EN_AA) & _BV(ENAA_P1)) + 48),
+            static_cast<char>(static_cast<bool>(read_register(EN_AA) & _BV(ENAA_P0)) + 48),
```
I assume this means that the auto-ack feature is always represented in binary format? This is another difference that should be noted in the docs for this function.

Originally, it should display `true`/`false` when all the pipes are uniformly configured.

dstroy0 replied:
I reenabled that behavior, the pointer points to the string literal "Enabled" if true else it prints out bitwise which I believe is the default behavior.

dstroy0 commented:
1. I'm using it on ESP32/ESP8266 it should be ok for anything with more than 2k of free ram. The strlen on my implementation is 869,
2. the formatting string literal adds to flash memory,  I thought about adding a directive toggle for the new members but I wasn't sure if I should do that, I'm new to using github and to contributing.
3. I wanted to be able to send the object over mqtt and I'm also able to print it out over the free version of webserial now
4. I'm learning how to use github and I'll work on the other stuff you asked about once I figure out how, probably tomorrow

2bndy5 commented:
Ran the CI tests, and this breaks all builds for the Pico SDK and most builds for Linux. Fortunately, I think its an easy fix: You need to add `#include <stdio.h>` to the RF24.cpp file. This is a common gotcha for people coming from Arduino land because `#include <Arduino.h>` brings a lot of other header files.

If for some reason adding `#include <stdio.h>` breaks Arduino builds, then the line needs to be added to the RF24_config.h file (for any platform that needs it). I can't seem to trigger the CI builds for Arduino (which tests various boards provided by the Arduino IDE) & PlatformIO (which tests Teensy boards). Usually, a [commit like this](https://github.com/nRF24/RF24/pull/821/commits/ebd1f2c72ab2437fd2292d4ae6dda3827b126c83) would trigger them, but I don't know what is going on.

>I thought about adding a directive toggle for the new members but I wasn't sure if I should do that

Typically we only do this when a feature cannot/shouldn't be supported by a certain platform. So, if `sprintf()` isn't supported by a certain chip, then we'd need to do this.

dstroy0 commented:
I tested on the mega2560 r3, working.

2bndy5 commented:
I'm more concerned as to why the CI workflows aren't getting triggered. This makes testing the changes on various boards and platforms super easy (without having to hookup a physical board and everything). I'm thinking I need to close and re-open this PR to trigger the CI... We need to make sure the `#include <stdio.h>` fixed the builds for the Pico SDK and Linux.

2bndy5 commented:
We tend to use doxygen primarily for documentation generation from src code comments. If you run `doxygen` at the root folder of this repo, then point you internet browser to the RF24/docs/html/classRF24.html file you should be able to see how the new function is documented.

I say all of this because the new function needs a:
```
@param debugging_information The c-string buffer that the debugging information is stored to.
```
You could word the description differently and elaborate how you like, but the parameter isn't documented for `void RF24::sprintfPrettyDetails(char * debugging_information)`

dstroy0 commented:
> I'm more concerned as to why the CI workflows aren't getting triggered. This makes testing the changes on various boards and platforms super easy (without having to hookup a physical board and everything). I'm thinking I need to close and re-open this PR to trigger the CI... We need to make sure the `#include <stdio.h>` fixed the builds for the Pico SDK and Linux.

I don't know what CI workflows are, I've got a lot to learn about contributing and collaborating.  AVR has their own stdio.h, I don't know if sprintf_P is standard C++.  If it's not there are other ways I can get around that.

2bndy5 commented:
CI stands for Continuous Integration. [Github has docs about using them](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwj94p_bo6_1AhXGI0QIHZ-eAJ4QFnoECAsQAQ&url=https%3A%2F%2Fdocs.github.com%2Fen%2Factions&usg=AOvVaw2m5bBnxqL_mvktdrLEIc5_), and the results tend to be in the repo's ["Actions" tab at the top of the page](https://github.com/nRF24/RF24/actions).

`sprintf()` is part the C standard lib stdio.h, and I believe it should also define the macro `sprintf_P` in that header file.

2bndy5 commented:
Just built this locally on ubuntu, and it turns out that we need
```c++
#include <stdio.h>

#ifndef sprintf_P
    #define sprintf_P sprintf
#endif
```
I'll have to get the pico SDK updated for my machine, and check that fixes the Pico SDK builds as well...

2bndy5 commented:
Ok, the `#ifndef sprintf_P` fix works for the Pico SDK builds as well.

BTW, it took almost 15 minutes to build the RPi Pico examples for 1 board. And that took place after the 5-10 minutes to setup the Pico SDK for my machine. The CI workflow builds the Pico examples for all supported boards in less than 5 minutes; this is another reason I'm concerned with the CI workflows not getting triggered.

2bndy5 commented on RF24.h:
```diff
@@ -598,6 +598,46 @@ class RF24 {
+     * hexadecimal to binary. Only use this function if your application can
+     * spare extra bytes of memory.
+     *
+     * @warning use a buffer of sufficient size for debugging_information, start 
+     * with a sizeof >= 1024
+     ^ @param debugging_information The c-string buffer that the debugging 
```
1. "sizeof" should be space delimited: "size of"
2. Please use proper English and capitalize the first word of the sentence: "Use a buffer ..." not "use a buffer...". It also wouldn't hurt to have a period at the end.
3. The `^` seems to be a typo again. The line should start with `*`. Just to show why this is important:
![image](https://user-images.githubusercontent.com/14963867/149393811-e516a4c6-4ef6-49e3-a807-afc0745768bd.png)
Notice the typo `^` shows up at the end of "size of >= 1024"

2bndy5 commented on RF24.h:
```diff
@@ -598,6 +598,46 @@ class RF24 {
+     ^ @param debugging_information The c-string buffer that the debugging 
+     *  information is stored to.
+     * 
+     * @note http://www.cplusplus.com/reference/cstdio/sprintf/
```
Please change `@note` to `@remark`. This is to
1. avoid doxygen drawing too much attention for otherwise basic C feature (like using `sprintf()`)
2. I don't want to pollute the doc with unnecessary admonished text.

This is not a mistake on your behalf. Its just me being very sensitive to the docs I've put so much work into.

dstroy0 replied:
I understand and made the requested changes.

2bndy5 reviewed (changes requested):
My second review: This is looking better. We are now down to fixing English stuff (& a syntax typo)

2bndy5 commented on RF24.h:
```diff
@@ -598,6 +598,46 @@ class RF24 {
+     *  information is stored to.
+     * 
+     * @note http://www.cplusplus.com/reference/cstdio/sprintf/
+     ^ sprintf_P formats a C-string in the exact same way as printf_P and outputs 
```
Another line that should begin with `*` not `^`. I'm curious if this is actually a typo or are you using a different keyboard (non-querty)?

dstroy0 replied:
I think it's brackets, which is what I'm using to edit these files with.  I'll pay close attention so I don't keep breaking things.

2bndy5 replied:
I highly recommend using vscode when/if you can. Even gitpod.io has started using vscode as their online IDE.

dstroy0 replied:
I will go check that out right now!

2bndy5 reviewed (approved):
Congrats @dstroy0 ! I can now accept this change into master. [Here is what the new function's docs look like on readthedocs,org.](https://rf24--821.org.readthedocs.build/en/821/classRF24.html#_CPPv4N4RF2420sprintfPrettyDetailsEPc)

@Avamander @TMRh20 Any objections/comments?

2bndy5 commented:
I fixed the indentation (manually) and trimmed the trailing whitespaces (a feature in vscode).

### about the python wrapper
Currently, this addition is not present in the python wrapper, and I added this fact to the function's docs.

I don't think that this function would be easy to add to the python wrapper as there would have to be a conversion from c-string obj to python unicoded string obj implemented when calling a wrapped/modified version of `sprintfPrettyDetails()` from the python interpreter. Using boost.python makes this more difficult than using pyind11. I suppose we can cross that bridge if someone makes a case for it...

dstroy0 commented:
> I fixed the indentation (manually) and trimmed the trailing whitespaces (a feature in vscode).
> ### about the python wrapper
> 
> Currently, this addition is not present in the python wrapper, and I added this fact to the function's docs.
> 
> I don't think that this function would be easy to add to the python wrapper as there would have to be a conversion from c-string obj to python unicoded string obj implemented when calling a wrapped/modified version of `sprintfPrettyDetails()` from the python interpreter. Using boost.python makes this more difficult than using pyind11. I suppose we can cross that bridge if someone makes a case for it...

I need to get better with python so I'll give it a try, I've never used the python wrapper.

2bndy5 commented:
>I need to get better with python so I'll give it a try, I've never used the python wrapper.

Its awesome you're willing to try! Here's some pointers:
1. source code for the python wrapper is written in C/C++ and can be found in RF24/pyRF24/pyRF24.cpp. So, you don't actually need to learn python syntax until you go to test out your newly wrapped function within a python interpreter/REPL.
2. We use boost.python to help wrap the RF24 class into a python interface. It might help to understand the build process (which is [outlined in the docs](https://rf24.readthedocs.io/en/latest/md_docs_python_wrapper.html)). The hardest part about this is reading [the boost.python docs](https://www.boost.org/doc/libs/1_78_0/libs/python/doc/html/index.html) and the [pydocs about creating C-extensions for CPython](https://docs.python.org/3/extending/extending.html).
3. The way I envision this in python would be like so:
   ```py
   from RF24 import RF24
   radio = RF24(22, 0)
   if not radio.begin():
       print("failed to init SPI")

   # debug_info is a str() that is returned from the wrapper
   debug_info = radio.sprintfPrettyDetails()
   print(debug_info)
   ```
   This way we don't have to <sup>1</sup>take in a python str as an arg, <sup>2</sup>convert it to a c-string to use assemble the info, and <sup>3</sup>convert it back to a python string when returning the buffer.
   We would simply <sup>1</sup>create a c-string with a wrapped version of the new function, <sup>2</sup>pass that c-string to the function you wrote already, and <sup>3</sup>return that c-string after converting it to a python string.

2bndy5 commented:
[This CPython function](https://docs.python.org/3/c-api/unicode.html#c.PyUnicode_FromStringAndSize) might make your life easier.

2bndy5 commented:
I decided not to port this feature to Circuitpython_nRF24L01 lib because <sup>1</sup>memory on Circuitpython running that lib is already being stretched thin for networking API and <sup>2</sup>there is already enough exposed API (including private members - because its pure python) for users to manage their own implementation of this feature.

@dstroy0 Mind if I take try to hack this feature into the python wrapper or are you still cooking up another surprise?

dstroy0 commented:
I’m still reading the boost python docs.  Should I keep this pull request open?  I’m curious to see the solution.  I’ve got vscode up and running, getting my feet wet using it.  Is there anything else that needs attention? 

> On Jan 14, 2022, at 2:29 PM, Brendan ***@***.***> wrote:
> 
> ﻿
> I decided not to port this feature to Circuitpython_nRF24L01 lib because 1memory on Circuitpython running that lib is already being stretched thin for networking API and 2there is already enough exposed API (including private members - because its pure python) for users to manage their own implementation of this feature.
> 
> @dstroy0 Mind if I take try to hack this feature into the python wrapper or are you still cooking up another surprise?
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android. 
> You are receiving this because you were mentioned.

2bndy5 commented:
Its not unusual for a PR to remain open while the feature it introduces gets ironed out. Aside from settling what to do with the python wrapper, I don't think anything else needs attention. I'll try coding what I've got stuck in my head and see if it builds. If it builds I'll commit it and test it on RPi...

## Words of the wisedom
- Most new features are hypothetically discussed in an issue before submitting a PR. This is the stage of this PR currently because no issue preceeded it.
- When submitting a PR, it usually originates from a non-main/non-master branch of the fork. [Github has docs about typical contributing workflows](https://docs.github.com/en/get-started/quickstart/github-flow). This can also be done in vscode instead of using github
   1. click the branch icon
       ![image](https://user-images.githubusercontent.com/14963867/149582536-d6208af3-fab9-45a5-9bc1-d96bde037bdd.png)
   2. select an existing branch or create a new one
       ![image](https://user-images.githubusercontent.com/14963867/149582738-0cc7f4ec-8249-43d8-a617-254053deb8cc.png)

2bndy5 commented:
Sometimes I surprise myself when I get it right the first time!

@dstroy0 This commit works to add the new function to the python wrapper. Result from testing on my RPi4:
```py
>>> from RF24 import RF24
>>> radio = RF24(22, 0)
>>> radio.begin()
True
>>> s = radio.sprintfPrettyDetails()
>>> print(s)
================ SPI Configuration ================
CSN Pin                 = 0
CE Pin                  = 22
SPI Frequency           = 10 Mhz
================ NRF Configuration ================
Channel                 = 76 (~ 2476 MHz)
RF Data Rate            = 1 MBPS
RF Power Amplifier      = PA_MAX
RF Low Noise Amplifier  = Enabled
CRC Length              = 16 bits
Address Length          = 5 bytes
Static Payload Length   = 32 bytes
Auto Retry Delay        = 1500 microseconds
Auto Retry Attempts     = 15 maximum
Packets lost on
    current channel     = 0
Retry attempts made for
    last transmission   = 0
Multicast               = Disabled
Custom ACK Payload      = Disabled
Dynamic Payloads        = Disabled
Auto Acknowledgment     = Enabled
Primary Mode            = TX
TX address              = 0x65646f4e32
pipe 0 (  open  ) bound = 0x65646f4e32
pipe 1 (  open  ) bound = 0x65646f4e31
pipe 2 ( closed ) bound = 0xc3
pipe 3 ( closed ) bound = 0xc4
pipe 4 ( closed ) bound = 0xc5
pipe 5 ( closed ) bound = 0xc6


>>>repr(s)
"'================ SPI Configuration ================\\nCSN Pin\\t\\t\\t= 0\\nCE Pin\\t\\t\\t= 22\\nSPI Frequency\\t\\t= 10 Mhz\\n================ NRF Configuration ================\\nChannel\\t\\t\\t= 76 (~ 2476 MHz)\\nRF Data Rate\\t\\t= 1 MBPS\\nRF Power Amplifier\\t= PA_MAX\\nRF Low Noise Amplifier\\t= Enabled\\nCRC Length\\t\\t= 16 bits\\nAddress Length\\t\\t= 5 bytes\\nStatic Payload Length\\t= 32 bytes\\nAuto Retry Delay\\t= 1500 microseconds\\nAuto Retry Attempts\\t= 15 maximum\\nPackets lost on\\n    current channel\\t= 0\\r\\nRetry attempts made for\\n    last transmission\\t= 0\\r\\nMulticast\\t\\t= Disabled\\nCustom ACK Payload\\t= Disabled\\nDynamic Payloads\\t= Disabled\\nAuto Acknowledgment\\t= Enabled\\nPrimary Mode\\t\\t= TX\\nTX address\\t\\t= 0x65646f4e32\\npipe 0 (  open  ) bound\\t= 0x65646f4e32\\npipe 1 (  open  ) bound\\t= 0x65646f4e31\\npipe 2 ( closed ) bound\\t= 0xc3\\npipe 3 ( closed ) bound\\t= 0xc4\\npipe 4 ( closed ) bound\\t= 0xc5\\npipe 5 ( closed ) bound\\t= 0xc6\\n\\n'"
```
However, notice that there is extra whitespace at the end of the returned string; executing `repr(s)` shows that the string ends with `\n\n`. While any call to `print()` will append a `\n` automatically, it seems that the returned value already ends in a `\n`. Is this expected behavior for arduino? I think that the user can control the last line's termination by using `Serial.println()` vs `Serial.print()`

2bndy5 commented on RF24.cpp:
```diff
@@ -783,6 +817,175 @@ void RF24::printPrettyDetails(void) {
+                                     PRIPSTR
+                                     " ) bound\t= 0x"
+                                     PRIPSTR
+                                     "\n\n"
```
I don't think this extra `\n\n` is really necessary as the user can control how the last line terminates.

2bndy5 replied:
we could probably also save an extra 12 bytes by getting rid of the padded spaces around the `closed`/` open ` status for each pipe.
Changing
```
pipe 0 (  open  ) bound = 0x65646f4e32
pipe 1 (  open  ) bound = 0x65646f4e31
pipe 2 ( closed ) bound = 0xc3
pipe 3 ( closed ) bound = 0xc4
pipe 4 ( closed ) bound = 0xc5
pipe 5 ( closed ) bound = 0xc6
```
into 
```
pipe 0 ( open ) bound = 0x65646f4e32
pipe 1 ( open ) bound = 0x65646f4e31
pipe 2 (closed) bound = 0xc3
pipe 3 (closed) bound = 0xc4
pipe 4 (closed) bound = 0xc5
pipe 5 (closed) bound = 0xc6
```

2bndy5 replied:
With the above changes, I calculate a maximum 861 bytes needed for any resulting output from `sprintfPrettyDetails()` This calculation includes
- binary representation of auto-ack
- LNA = "Disabled"
- Channel > 99
- max values of "Packets lost ..." & "Retry attempts made ..."

2bndy5 replied:
@dstroy0 If your scenario is ok with these changes, then I'll commit them. I'm not sure if the `\n\n` is required for sending the c-string over MQTT, but I think it can still be added on the user's end not the lib's end.

dstroy0 commented:
It’s not necessary, you are correct the user should specify spacing after.  Just the one new line makes it like the output of printPrettyDetails.  Any padding you want to remove to make the output smaller is ok with me.  The function local buffer sizes can be optimized too so that memory use for the function goes down.  

> On Jan 17, 2022, at 2:11 PM, Brendan ***@***.***> wrote:
> 
> ﻿
> @2bndy5 commented on this pull request.
> 
> In RF24.cpp:
> 
> > +                                     "\n"
> +                                     "pipe 3 ( "
> +                                     PRIPSTR
> +                                     " ) bound\t= 0x"
> +                                     PRIPSTR
> +                                     "\n"
> +                                     "pipe 4 ( "
> +                                     PRIPSTR
> +                                     " ) bound\t= 0x"
> +                                     PRIPSTR
> +                                     "\n"
> +                                     "pipe 5 ( "
> +                                     PRIPSTR
> +                                     " ) bound\t= 0x"
> +                                     PRIPSTR
> +                                     "\n\n"
> @dstroy0 If your scenario is ok with these changes, then I'll commit them. I'm not sure if the \n\n is required for sending the c-string over MQTT, but I think it can still be added on the user's end not the lib's end.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android. 
> You are receiving this because you were mentioned.

2bndy5 commented:
I'm ready to merge this now.
python wrapper test:
```py
>>> s = radio.sprintfPrettyDetails()
>>> print(s)
================ SPI Configuration ================
CSN Pin                 = 0
CE Pin                  = 22
SPI Frequency           = 10 Mhz
================ NRF Configuration ================
Channel                 = 76 (~ 2476 MHz)
RF Data Rate            = 1 MBPS
RF Power Amplifier      = PA_MAX
RF Low Noise Amplifier  = Enabled
CRC Length              = 16 bits
Address Length          = 5 bytes
Static Payload Length   = 32 bytes
Auto Retry Delay        = 1500 microseconds
Auto Retry Attempts     = 15 maximum
Packets lost on
    current channel     = 0
Retry attempts made for
    last transmission   = 0
Multicast               = Disabled
Custom ACK Payload      = Disabled
Dynamic Payloads        = Disabled
Auto Acknowledgment     = Enabled
Primary Mode            = TX
TX address              = 0x65646f4e32
pipe 0 ( open ) bound   = 0x65646f4e32
pipe 1 ( open ) bound   = 0x65646f4e31
pipe 2 (closed) bound   = 0xc3
pipe 3 (closed) bound   = 0xc4
pipe 4 (closed) bound   = 0xc5
pipe 5 (closed) bound   = 0xc6
>>> len(s)
856
```
Any more optimizations can be submitted via another PR. I just want to get this upstream because it may well support boards that can't use `printDetails()` due to lack of `printf()` support.

2bndy5 commented:
well the CI workflows finally got triggered...

@dstroy0 You can see that this change does not harm builds for any of the officially supported boards in the [Arduino CI run](https://github.com/nRF24/RF24/actions/runs/1710056860). The [PlatformIO CI run](https://github.com/nRF24/RF24/actions/runs/1710056851) shows all green as well.

Keep in mind that these workflows only test compile all the examples, and none of the examples actually call `sprintfPrettyDetails()`. So, it may not be the best approach for code coverage, but at least we know that the next release won't be completely broken.

dstroy0 commented:
Excellent!

> On Jan 17, 2022, at 4:48 PM, Brendan ***@***.***> wrote:
> 
> ﻿
> I'm ready to merge this now.
> python wrapper test:
> 
> >>> s = radio.sprintfPrettyDetails()
> >>> print(s)
> ================ SPI Configuration ================
> CSN Pin                 = 0
> CE Pin                  = 22
> SPI Frequency           = 10 Mhz
> ================ NRF Configuration ================
> Channel                 = 76 (~ 2476 MHz)
> RF Data Rate            = 1 MBPS
> RF Power Amplifier      = PA_MAX
> RF Low Noise Amplifier  = Enabled
> CRC Length              = 16 bits
> Address Length          = 5 bytes
> Static Payload Length   = 32 bytes
> Auto Retry Delay        = 1500 microseconds
> Auto Retry Attempts     = 15 maximum
> Packets lost on
>     current channel     = 0
> Retry attempts made for
>     last transmission   = 0
> Multicast               = Disabled
> Custom ACK Payload      = Disabled
> Dynamic Payloads        = Disabled
> Auto Acknowledgment     = Enabled
> Primary Mode            = TX
> TX address              = 0x65646f4e32
> pipe 0 ( open ) bound   = 0x65646f4e32
> pipe 1 ( open ) bound   = 0x65646f4e31
> pipe 2 (closed) bound   = 0xc3
> pipe 3 (closed) bound   = 0xc4
> pipe 4 (closed) bound   = 0xc5
> pipe 5 (closed) bound   = 0xc6
> >>> len(s)
> 856
> Any more optimizations can be submitted via another PR. I just want to get this upstream because it may well support boards that can't use printDetails() due to lack of printf() support.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android. 
> You are receiving this because you were mentioned.
